apis/net/ieee802154/src/rx.rs

```diff
@@ -0,0 +1,310 @@
+
+            // Safety: kernel schedules an upcall iff a new frame becomes available,
+            // i.e. when it increments `read_index`.
+            Ok(self.buf.next_frame())
+        }
+    }
+}
+
+// Reception
+impl<S: Syscalls, C: Config> Ieee802154<S, C> {
+    fn receive_frame_single_buf<const N: usize>(
+        buf: &mut RxRingBuffer<N>,
+    ) -> Result<(), ErrorCode> {
+        let called: Cell<Option<(u32,)>> = Cell::new(None);
+        share::scope::<
+            (
+                AllowRw<_, DRIVER_NUM, { allow_rw::READ }>,
+                Subscribe<_, DRIVER_NUM, { subscribe::FRAME_RECEIVED }>,
+            ),
+            _,
+            _,
+        >(|handle| {
+            let (allow_rw, subscribe) = handle.split();
+            S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(allow_rw, buf.as_mut_byte_slice())?;
+            S::subscribe::<_, _, C, DRIVER_NUM, { subscribe::FRAME_RECEIVED }>(subscribe, &called)?;
+
+            loop {
+                S::yield_wait();
+                if let Some((_lqi,)) = called.get() {
+                    // At least one frame was received.
+                    return Ok(());
+                }
+            }
+        })
+    }
+}
+
+// This module is to protect RxRingBufferInKernel private API.
+mod alternate_receive_buffers {
+    use super::*;
+
+    pub(super) struct RxRingBufferInKernel<'buf, const N: usize, S: Syscalls, C: Config> {
+        lent_buf: *mut RxRingBuffer<N>,
+        covariance_phantom: PhantomData<&'buf mut RxRingBuffer<N>>,
+        sc: PhantomData<(S, C)>,
+    }
+
+    impl<'buf, const N: usize, S: Syscalls, C: Config> Drop for RxRingBufferInKernel<'buf, N, S, C> {
+        // Unshares the buffer to prevent kernel accessing no longer valid memory.
+        fn drop(&mut self) {
+            share::scope::<AllowRw<_, DRIVER_NUM, { allow_rw::READ }>, _, _>(|handle| {
+                let _ = S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(handle, &mut []);
+            });
+        }
+    }
+
+    impl<'buf, const N: usize, S: Syscalls, C: Config> RxRingBufferInKernel<'buf, N, S, C> {
+        fn new(buf: &mut RxRingBuffer<N>) -> Self {
+            Self {
+                lent_buf: buf as *mut RxRingBuffer<N>,
+                covariance_phantom: PhantomData,
+                sc: PhantomData,
+            }
+        }
+
+        pub(super) fn share_initial(buf: &mut RxRingBuffer<N>) -> Result<Self, ErrorCode> {
+            Self::share_unscoped(buf.as_mut_byte_slice())?;
+            Ok(Self::new(buf))
+        }
+
+        fn share_unscoped(buf: &mut [u8]) -> Result<(), ErrorCode> {
+            let allow_rw = platform::AllowRw::<S, DRIVER_NUM, { allow_rw::READ }>::default();
+
+            // Safety: The buffer being allowed here is going to be enclosed in an opaque type
+            // until it's unallowed again. This prevents concurrent access to the buffer by process and kernel.
+            let allow_rw_handle = unsafe { share::Handle::new(&allow_rw) };
```

Comment:
`Handle::new`'s safety invariant is:

```rust
impl<'handle, L: List> Handle<'handle, L> {
    /// Constructs a new `Handle`, typing its lifetime to the provided list.
    ///
    /// # Safety
    /// The calling code must guarantee that `Drop::drop` is called on `_list`'s
    /// pointee before `_list`'s pointee becomes invalid. In other words,
    /// `_list`'s pointee may not be forgotten or leaked.
    pub unsafe fn new(_list: &'handle L) -> Self {
        Handle { _list: PhantomData }
    }
```

In particular, it says "`_list`'s pointee may not be forgotten or leaked". In this case, `_list` is `&allow_rw`, so its pointee is `allow_rw`, which is explicitly forgotten later in the function (`core::mem::forget(allow_rw)`).

`share_unscoped` does not follow this requirement, making it unsound. In general: `share_unscoped` does not do anything to guarantee that the buffer will be unallowed before it is used again by the application.

Reply:
> In general: `share_unscoped` does not do anything to guarantee that the buffer will be unallowed before it is used again by the application.

That's true. It's only intended to be used as a helper function, not as a safe API. If I mark it `unsafe`, will it be enough for it?

> `Handle::new`'s safety invariant is:
> 
> ```rust
> impl<'handle, L: List> Handle<'handle, L> {
>     /// Constructs a new `Handle`, typing its lifetime to the provided list.
>     ///
>     /// # Safety
>     /// The calling code must guarantee that `Drop::drop` is called on `_list`'s
>     /// pointee before `_list`'s pointee becomes invalid. In other words,
>     /// `_list`'s pointee may not be forgotten or leaked.
>     pub unsafe fn new(_list: &'handle L) -> Self {
>         Handle { _list: PhantomData }
>     }
> ```
> 
> In particular, it says "`_list`'s pointee may not be forgotten or leaked". In this case, `_list` is `&allow_rw`, so its pointee is `allow_rw`, which is explicitly forgotten later in the function (`core::mem::forget(allow_rw)`).
> 
> `share_unscoped` does not follow this requirement, making it unsound.

OK, `share_unscoped` itself allow memory violation (concurrent mutation). But again, it's not public safe API. The idea is that `share_initial` is sound, because:
1. as long as `RxRingBufferInKernel` exists, it quasi-borrows the buffer (by means of `covariance_phantom`), thus preventing accesing the buffer from userspace;
2. when `RxRingBufferInKernel` is dropped, kernel is un-allowed the buffer, so now only userspace can access the buffer.

Why isn't the above true? I can easily convince myself this is sound, and I'm sad I can't convince Rust about the same.

Reply:
I suspect there's a relatively-easy fix to avoid the aliasing issue, but unfortunately I don't think there is an easy fix to make `RxRingBufferInKernel` sound.

A buffer that has been shared with the kernel via Read-Write Allow MUST be un-Allowed before its memory is deallocated. `RxRingBufferInKernel` does not currently guarantee this: if the caller calls `core::mem::forget` on it, then `Drop::drop` will never be called, and the shared buffer will be deallocated without being first un-Allowed (resulting in UB).

I'm only aware of a handful of ways to write a sound API that uses Read-Write Allow:

1. Use the same pattern as `share::scope` (have a function call a closure and guarantee that the buffer is un-allowed when the closure returns).
2. Use [`Pin`](https://github.com/tock/libtock-rs/issues/494#issuecomment-1851077413).
3. Require the buffer to be `'static`.
4. Use dynamic memory allocation (note: we don't currently have dynamic memory allocation implemented in libtock-rs, but we want to have it as an option).

Reply:
> I suspect there's a relatively-easy fix to avoid the aliasing issue, but unfortunately I don't think there is an easy fix to make RxRingBufferInKernel sound.

@jrvanwhy Could you please share the idea how to avoid the aliasing issue?